### PR TITLE
Require reducers to define a monoid

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/pull/6065#issuecomment-1518014632.

### CODE REVIEW

**Review by fnrizzi — APPROVED** [review]

File: docs/source/API/core/builtinreducers/ReducerConcept.rst [hunk 1]
```diff
@@ -92,6 +92,22 @@ Functions
...
+
+.. code-block:: cpp
+
+    value_type op(const value_type& val1, const value_type& val2) {
+      value_type result = val1;
+      reducer.join(result, val2);
+      return result;
+    }
+
+is commutative and associative with identity element defined by ``reducer.init(el)``.
```
Comment (dalg24) [inline docs/source/API/core/builtinreducers/ReducerConcept.rst:108]: > with identity element defined by ``reducer.init(el)``.

Sounds a bit weird to me but not sure how to improve.

Comment (fnrizzi) [reply docs/source/API/core/builtinreducers/ReducerConcept.rst:108]: per discussion: with identity element that can be set by calling ``reducer.init(el)``.

Comment (masterleinad) [reply docs/source/API/core/builtinreducers/ReducerConcept.rst:108]:
```suggestion
is commutative and associative with identity element that can be set by calling ``reducer.init(el)``.
```

**Review by crtrott — APPROVED** [review]

File: docs/source/API/core/builtinreducers/ReducerConcept.rst [hunk 2]
```diff
@@ -92,6 +92,22 @@ Functions
...
+.. code-block:: cpp
+
+    value_type op(const value_type& val1, const value_type& val2) {
+      value_type result = val1;
+      reducer.join(result, val2);
+      return result;
+    }
+
+is commutative and associative with identity element that can be set by calling ``reducer.init(el)``.
+
```
Comment (crtrott) [inline docs/source/API/core/builtinreducers/ReducerConcept.rst:109]: We could also go with a "non-mathy" explanation:
```c++
auto foo(auto reducer, auto val1, auto val2) {
   auto result = val1;
   reducer.join(result, val2);
   return result;
}
```

returns `val` for both `foo(reducer, init, val)` and `foo(reducer, val, init)` where `init` is a value set by calling `reducer.init(init)`.

Comment (dalg24) [reply docs/source/API/core/builtinreducers/ReducerConcept.rst:109]: I think your suggestion is a bit more appropriate for the wiki.

Comment (dalg24) [reply docs/source/API/core/builtinreducers/ReducerConcept.rst:109]: Do you want to open a PR for it?

**Pull request merged**